windows_integration/client/src/convert.rs

```diff
@@ -0,0 +1,29 @@
+use tracing::{event, Level};
+use windows::{Win32::Foundation::UNICODE_STRING, core::PWSTR};
+
+pub fn unicode_to_rust(string: UNICODE_STRING) -> Option<String> {
+	match unsafe { string.Buffer.to_string() } {
```

Comment:
This is unsafe, and has buffer overrun problems: https://github.com/microsoft/windows-rs/issues/2566

Reply:
@micolous After reading the issue I believe this is fixed with 1954fb05cf7b9217708f86565c3690826f24cdba

Reply:
It's not, calling `Buffer.to_string()` is unsafe because it uses `wcslen` to determine the length of the string rather than the `Length` field.

You need to access the raw `*mut u16` pointer from inside the `PWSTR` (`Buffer.0`), and form a slice using the `Length` parameter:

```rust
if string.Length == 0 {
  // MSDN docs state Buffer could be empty for zero length
  return Some(String::new());
}
let result = String::from_utf16(unsafe { std::slice::from_raw_parts(string.Buffer.0, string.Length.into() / 2) });

// error handling for UTF-16 decoding
// ...
```

You should also check:
* `string.Buffer.0` isn't null, and it's properly ali